MatrixSDK/Utils/MXUIKitBackgroundTask.m

```diff
@@ -46,39 +54,71 @@ - (instancetype)initWithName:(NSString*)name expirationHandler:(MXBackgroundTask
     return self;
 }
 
-
-- (instancetype)initAndStartWithName:(NSString*)name expirationHandler:(MXBackgroundTaskExpirationHandler)expirationHandler
+- (instancetype)initAndStartWithName:(NSString*)name
+                   expirationHandler:(MXBackgroundTaskExpirationHandler)expirationHandler
 {
-    self = [super init];
+    self = [self initWithName:name expirationHandler:expirationHandler];
     if (self)
     {
-        self.name = name;
-        
         UIApplication *sharedApplication = [self sharedApplication];
         if (sharedApplication)
         {
-            self.startDate = [NSDate date];
-            
             MXWeakify(self);
             
-            self.identifier = [sharedApplication beginBackgroundTaskWithName:self.name expirationHandler:^{
-                
+            // Note: -[UIApplication applicationState] must be used from main thread only
+            dispatch_async(dispatch_get_main_queue(), ^{
                 MXStrongifyAndReturnIfNil(self);
                 
-                NSLog(@"[MXBackgroundTask] Background task expired #%lu - %@ after %.0fms", (unsigned long)self.identifier, self.name, self.elapsedTime);
+                //  application is in background and not enough time to start this task
+                if (sharedApplication.applicationState == UIApplicationStateBackground &&
```

Comment:
Can be interesting to make the threshold configurable in the future

Reply:
Indeed.